components/filecoin/rs/src/signature.rs

```diff
@@ -0,0 +1,187 @@
+use crate::message::MessageAPI;
+use blake2b_simd::Params;
+use bls_signatures::Serialize;
+use core::{array::TryFromSliceError, num::ParseIntError};
+use fvm_ipld_encoding::Cbor;
+use fvm_shared::address::Protocol;
+use fvm_shared::message::Message as UnsignedMessage;
+use libsecp256k1::util::{SECRET_KEY_SIZE, SIGNATURE_SIZE};
+use libsecp256k1::{sign, Message};
+use std::convert::TryInto;
+use thiserror::Error;
+
+pub const SIGNATURE_RECOVERY_SIZE: usize = SIGNATURE_SIZE + 1;
+pub const BLS_SIGNATURE_SIZE: usize = 96;
+
+pub struct PrivateKey(pub [u8; SECRET_KEY_SIZE]);
+pub struct SignatureBLS(pub [u8; BLS_SIGNATURE_SIZE]);
+pub struct SignatureSECP256K1(pub [u8; SIGNATURE_RECOVERY_SIZE]);
+
+pub enum Signature {
+    SignatureSECP256K1(SignatureSECP256K1),
+    SignatureBLS(SignatureBLS),
+}
+
+impl Signature {
+    pub fn as_bytes(&self) -> &[u8] {
+        match self {
+            Signature::SignatureSECP256K1(sig_secp256k1) => sig_secp256k1.as_bytes(),
+            Signature::SignatureBLS(sig_bls) => sig_bls.as_bytes(),
+        }
+    }
+}
+
+impl SignatureBLS {
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl SignatureSECP256K1 {
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+/// Filecoin Signer Error
+#[derive(Error, Debug)]
+pub enum SignerError {
+    ///  CBOR error
+    #[error("CBOR error: '{0}'")]
+    CBOR(#[from] serde_cbor::Error),
+    /// Secp256k1 error
+    #[error("secp256k1 error")]
+    Secp256k1(#[from] libsecp256k1::Error),
+    // Key decoding error
+    #[error("key decoding error (only hex or base64 is accepted)")]
+    KeyDecoding(),
+    /// Hex Error
+    #[error("Hex decoding error | {0}")]
+    HexDecode(#[from] hex::FromHexError),
+    /// InvalidBigInt error
+    #[error("InvalidBigInt error")]
+    InvalidBigInt(#[from] num_bigint_chainsafe::ParseBigIntError),
+    /// Generic error message
+    #[error("Error: `{0}`")]
+    GenericString(String),
+    /// Not able to parse integer
+    #[error("Cannot parse integer")]
+    ParseIntError(#[from] ParseIntError),
+    /// BLS error
+    #[error("BLS error | {0}")]
+    BLS(#[from] bls_signatures::Error),
+    /// Invalid BIP44Path
+    #[error("Invalid BIP44 path : `{0}`")]
+    InvalidBIP44Path(#[from] zx_bip44::errors::BIP44PathError),
+    /// BLS error
+    #[error("Couldn't convert from slice")]
+    TryFromSlice(#[from] TryFromSliceError),
+    /// Base64 decode Error
+    #[error("Base64 decode error | {0}")]
+    DecodeError(#[from] base64::DecodeError),
+
+    #[error("Marshall error | {0}")]
+    FvmSharedEncodingError(#[from] fvm_ipld_encoding::Error),
+}
+
+const CID_PREFIX: &[u8] = &[0x01, 0x71, 0xa0, 0xe4, 0x02, 0x20];
+const HASH_LENGTH: usize = 32;
+
+/// transform a message into a hashed message ready to be signed and following Filecoin standard
+pub fn get_digest(message: &[u8]) -> Result<[u8; HASH_LENGTH], TryFromSliceError> {
```

Comment:
Should we link to where these functions are from?
Also, I'm not sure about this, but should we add their license into credits in this case? cc @fmarier

Reply:
https://github.com/Zondax/filecoin-signing-tools/blob/2e7b005b4733761de11d4a252cb481ca5aedb029/signer/src/utils.rs#L19

Reply:
I meant put in the comment, same for other copied functions.

Reply:
done

Reply:
If you could put Zondax license (see https://github.com/Zondax/filecoin-signing-tools/blob/master/signer/LICENSE#L189-L201) in the file as a comment that should be enough.